lib/live_cluster/client/msgpack.py

```diff
@@ -0,0 +1,107 @@
+import struct
+from typing import Union
+from msgpack.fallback import Packer
+from msgpack import ExtType
+
+from lib.live_cluster.client.ctx import ASValue, ASValues, CDTContext, CTXItem, CTXItems
+
+AS_BYTES_STRING = 3
+AS_BYTES_BLOB = 4
+AS_BYTES_GEOJSON = 23
+ASVAL_CMP_EXT_TYPE = 0xFF
+ASVAL_CMP_WILDCARD = 0x00
+
+
+class CTXItemWireType:
+    AS_CDT_CTX_LIST_INDEX = 0x10
+    AS_CDT_CTX_LIST_RANK = 0x11
+    AS_CDT_CTX_LIST_VALUE = 0x13
+    AS_CDT_CTX_MAP_INDEX = 0x20
+    AS_CDT_CTX_MAP_RANK = 0x21
+    AS_CDT_CTX_MAP_KEY = 0x22
+    AS_CDT_CTX_MAP_VALUE = 0x23
+
+
+class ASPacker(Packer):
+    def __init__(self, autoreset=False):
+        super().__init__(autoreset=autoreset)
+
+    def pack(self, obj):
+        if isinstance(obj, ASValue):
+            self._pack_as_value(obj)
+            return
+        elif isinstance(obj, CDTContext):
+            self._pack_as_cdt_ctx(obj)
+            return
+
+        super().pack(obj)
```

Comment:
what gets handled by this default case?

Reply:
All msgpack native types.  Almost all roads lead to this base case since self.pack() is recursive.